src/main/java/org/apache/maven/plugins/toolchain/jdk/ToolchainDiscoverer.java

```diff
@@ -411,10 +411,12 @@ private Set<Path> findJdks() {
             installedDirs.add(Paths.get(userHome, "Library/Java/JavaVirtualMachines"));
         } else if (win) {
             installedDirs.add(Paths.get("C:\\Program Files\\Java\\"));
-            Path scoop = Paths.get(userHome, "scoop/apps");
+            Path scoop = Paths.get(userHome, "scoop", "apps");
             if (Files.isDirectory(scoop)) {
                 try (Stream<Path> stream = Files.list(scoop)) {
-                    stream.forEach(dirsToTest::add);
+                    // Scoop can install multiple versions of a Java distribution, we only take the one that is
+                    // currently selected.
+                    stream.map(path -> Paths.get(path.toString(), "current")).forEach(dirsToTest::add);
```

Comment:
The goal is to find all JDK installed, not only the `current` one.

My undersanding is that they are installed by scoop in `C:\Users\M67B014\scoop\apps\[jdk-name]\[version]`.
The `installedDirs` should then point to `C:\Users\M67B014\scoop\apps\[jdk-name]` because it's later scanned for JDK in its child directories, see 
https://github.com/apache/maven-toolchains-plugin/blob/0e858055930cdd3e6687c61e421978539a947dd3/src/main/java/org/apache/maven/plugins/toolchain/jdk/ToolchainDiscoverer.java#L432-L436
So by adding all direct children of `scoop\apps`, we should later have the children of those directories.

Are you sure the fix in line 414 is not sufficient ?

Reply:
I'll double-check again one of these days, but I'm pretty sure that `<userHome>\scoop\apps` is not enough. Scoop-installed Java runtimes reside under `<userHome>\scoop\apps\<packageName>\<version>`, with `<userHome>\scoop\apps\<packageName>\current` as a symlink to the currently active version of that package.

Regardless of that, the question remains: given that `openjdk21` has three installed versions (21.0.0, 21.0.1 and 21.0.2), do we want all of them to be detected or only the current one?

Reply:
> I'll double-check again one of these days, but I'm pretty sure that `<userHome>\scoop\apps` is not enough. Scoop-installed Java runtimes reside under `<userHome>\scoop\apps\<packageName>\<version>`

I think the code was already checking those.  `installedDirs` should contain all `<userHome>\scoop\apps\<packageName>` while `dirsToTest` should later contain all `<userHome>\scoop\apps\<packageName>\<version>`.

>, with `<userHome>\scoop\apps\<packageName>\current` as a symlink to the currently active version of that package.

That one will be ignored, as symlinks are always transformed to a canonical path.  That's the one stored in the toolchain.

> Regardless of that, the question remains: given that `openjdk21` has three installed versions (21.0.0, 21.0.1 and 21.0.2), do we want all of them to be detected or only the current one?

The code is currently written to get *all* JDK.  The notion of `current` is specific to a given installer and we not be the latest one.  We have heuristic to choose amongst multiple versions and the goal in the discovery phase is to escape from "just the one configured in the shell".  So I'd rather have them all listed here and choose during toolchain matching / selection.

Reply:
Please see [this gist](https://gist.github.com/mthmulders/3c13223127254ace12190b9927940ee3). To create it, I've added some debug logging to print which directories indeed contain `bin/javac` or `bin/javac.exe`. I did that after the `filter` in line 444 to not print all packages I have installed that aren't a Java runtime.

If we only look at `Paths.get(userHome, "scoop", "apps");`, we will miss all installed runtimes - only the "current" one (which `JAVA_HOME` points to) will be detected. This is [file-inspect-scoop-apps-packagename-txt](https://gist.github.com/mthmulders/3c13223127254ace12190b9927940ee3#file-inspect-scoop-apps-packagename-txt).

Looking at the `current` directory under each installed package will make the code detect the current version of each installed Java runtime. So that is a lot more, but still not all. This is [inspect-scoop-apps-packagename-current-txt](https://gist.github.com/mthmulders/3c13223127254ace12190b9927940ee3#file-inspect-scoop-apps-packagename-current-txt).

To find all of them (as you wrote), we'd need to look at each subdirectory of an installed package. It will incur a lot of disk scanning, because it will inspect each installed version of each Scoop package, but it will even detect multiple versions of the same Java runtime (see openjdk21, microsoft17-jdk). This is [inspect-scoop-apps-packagename-all-subdirs-txt](https://gist.github.com/mthmulders/3c13223127254ace12190b9927940ee3#file-inspect-scoop-apps-packagename-all-subdirs-txt). It will also report the *current* version of that package (so that's a duplicate), but we could filter those out I believe.